### ISSUE TITLE
Make it possible to select window by keyboard click

### ISSUE THREAD

**Issue by baest** [lwouis/alt-tab-macos#210]
It would be very nice to be able to select the window to focus by using the other keys on the keyboard. So alt-tab and then the windows appears with a character in front of it (like the Desktop number). When you press that character (I'm thinking a-z, 0-9) then the window opens. If you have more windows than 35 I don't know if there should be multiple characters or just no character for the last ones.

**Comment by lwouis** [comment]
Hi @baest! Thank you for sharing this idea!

There is a similar ticket with #178. Before anyone spends time implementing such a feature, I think it is important that there is a more developed specification for it. You mentioned for example the shortcut would be a-z, 0-9. Is this really a good mapping on the keyboard, when you left hand is already holding a modifier key, probably with your thumb? Look at RTS games like Starcraft 2. They use shortcuts close to the left hand like number, for control groups.

Also what happens for conflicting shortcuts, let's say if you set your main alt-tab shortcut to `alt-1` already?

I think all these questions and more should be answered now, so that we get a clearer picture of:

* Is the feature really a good idea? Use-cases, ergonomics, etc
* How much work is involved? Who's gonna take on the work, can we split it into installments, etc
* What are the interactions with existing features? How could this merge with #178 for example?
* Edge-cases

**Comment by SoPat712** [comment]
This doesn't seem like a very logical issue, and mouse hover seems like it was the best solution. I think this should just be marked closed, to be honest.

**Comment by lwouis** [comment]
I think @baest suggestion is interesting. Are you sure you understand the scenario? They basically would like hardcoded shortcuts like `alt+1` always focuses the first window, `alt+2` the second, etc. It's a very popular feature of other similar apps so I don't think we should discard it.

I think it could be done in a lightweight way by adding a checkbox `Number keys` here:

![image](https://user-images.githubusercontent.com/106195/92389200-9bff4980-f153-11ea-983d-62abc73d0dcb.png)

A deeper implementation would maybe allow more than 1 to 9, customize windows, etc. @baest What do you think?

**Comment by SoPat712** [comment]
I think it would personally be a feature that I would never use, but I could see it being useful. I have too many windows open at a time for that to work well, however.

**Comment by baest** [comment]
@lwouis I think that could be a good solution. You had good points for my original suggestion and I have to admit I didn't consider them before posting this issue. I'm always a fan of keyboard shortcuts and when AltTab is in focus I guess we could bind whatever keys to it. I think the lightweight checkbox suggestion is a very good start, might be more than enough and if popular, but people need more, then I guess new issues would be raised and the feature expanded if there is enough tuits.

**Comment by justinnoel** [comment]
I added this comment in #1082:

> Place a hotkey like "a", "b", "c", etc at the bottom right of each app/window that will let me just press that key to activate that app. Command Tab Plus does this with numbers, but that doesn't work well with lots of apps in the switcher.

Sample screenshot:

![Screenshot of macOS app switcher using Command Tab Plus. Each app has a number selector for quick switching.](https://user-images.githubusercontent.com/81643826/137350130-ef69cff5-d005-400a-847b-5e75dd134213.jpg)


I already use an app that does exactly this. It's not awkward at all.

Scenario:

1) You use the keyboard shortcut to trigger the AltTab switcher (⌘ ⇥) in my case.

2) AltTab switcher opens with all the apps displaying a letter like "a", "b", "c", etc. 

3) Let go of the second portion of the hot key (⇥ in my case) while still holding down the first key in the sequence (⌘ in my case) and then tap the "h" key to open the app that has the "h" label. There is no need for another modifier key.

I do this every day with Command Tab Plus with numbers. The problem is that there are only numbers 1 through 0 so the available quick selections is way too limited.


====

@SoPat712 : I purposefully disable the "Mouse hover" option. So often in the regular macOS switcher, my mouse happens to be in a spot hovered over an app and I decide NOT to switch so I let go of the keys. BOOM! Because I was "hovered" over one of the apps in the switcher, macOS switches to it.

In Command Tab Plus and now in AltTab, I've disabled that. "Hover" is not a solution to this feature request.

**Comment by xbladesub** [comment]
> I added this comment in #1082:
> 
> > Place a hotkey like "a", "b", "c", etc at the bottom right of each app/window that will let me just press that key to activate that app. Command Tab Plus does this with numbers, but that doesn't work well with lots of apps in the switcher.
> 
> Sample screenshot:
> 
> ![Screenshot of macOS app switcher using Command Tab Plus. Each app has a number selector for quick switching.](https://user-images.githubusercontent.com/81643826/137350130-ef69cff5-d005-400a-847b-5e75dd134213.jpg)
> 
> I already use an app that does exactly this. It's not awkward at all.
> 
> Scenario:
> 
> 1. You use the keyboard shortcut to trigger the AltTab switcher (⌘ ⇥) in my case.
> 2. AltTab switcher opens with all the apps displaying a letter like "a", "b", "c", etc.
> 3. Let go of the second portion of the hot key (⇥ in my case) while still holding down the first key in the sequence (⌘ in my case) and then tap the "h" key to open the app that has the "h" label. There is no need for another modifier key.
> 
> I do this every day with Command Tab Plus with numbers. The problem is that there are only numbers 1 through 0 so the available quick selections is way too limited.
> 
> ====
> 
> @SoPat712 : I purposefully disable the "Mouse hover" option. So often in the regular macOS switcher, my mouse happens to be in a spot hovered over an app and I decide NOT to switch so I let go of the keys. BOOM! Because I was "hovered" over one of the apps in the switcher, macOS switches to it.
> 
> In Command Tab Plus and now in AltTab, I've disabled that. "Hover" is not a solution to this feature request.

try dcmd.app. The best! (only 4$) worth it

**Comment by muescha** [comment]
i would like to see shortcuts for example the first letter of the app following by a number of a tab - or  allow to search for a app name / title
(for this mode the shortcuts for close/minimize/... maybe are disabled)

**Comment by thimplicity** [comment]
I have read through this thread, because I opened a duplicate of this question. Is this still under consideration?

**Comment by lwouis** [comment]
Anyone can decide to step in and contribute this feature. Please see my message above for required next steps: https://github.com/lwouis/alt-tab-macos/issues/210#issuecomment-608483487

**Comment by lwouis** [comment]
Interesting idea of temporaly pinned windows: https://github.com/lwouis/alt-tab-macos/issues/3425#issuecomment-2558396174

**Comment by randalltrini** [comment]
I am not sure if this suggestion is still open but I'd like to contribute how it can work for me (a non programmer) to improve workflow.

Currently I have alt tab mapped to a key and  trackpad gesture using BTT. When triggered, alttab stays open and then I cann look. for the window and either manually click OR use arrow keys and spacebar to select. 

However, with the amount of windows I many have open with projects, a visible number overlay to me would speed up my workflow. It is faster for my eyes to pick out the window and short cut in a grid of let's say 10 windows and select a shortcut, than use arrow keys to navigate or move the mouse to the preffered window. Perhaps using the same alt key as a modifier key plus a number would solve issues with window numbers over 9 (OR maybe a delay for pressing two numbers in succession when the user wants a window over 9 - just a suggestion, I am not a programmer, so not sure how hard to implement).

Just my 2c. Feel free to seek clarification from me, but I think this would be a killer feature when people use it (picture it: press ***alt tab***, then ***5*** and your window appears OR ***gesture trigger*** then ***4*** for same).

RD

**Comment by muhamedkarajic** [comment]
I like the alt + 1 etc, it makes it sound like my solution just differently approached. I think then my suggestions arent necessary, I would be happy with that.

**Comment by muhamedkarajic** [comment]
> Anyone can decide to step in and contribute this feature. Please see my message above for required next steps: [#210 (comment)](https://github.com/lwouis/alt-tab-macos/issues/210#issuecomment-608483487). Essentially someone needs to write comprehensive specs for this feature, draw mockups, then we can move on to implementing it.

I might do it in the next months when I have some spare time. Thx!

**Comment by AbdelrahmanHafez** [comment]
Hi @lwouis, and everyone. Thanks for maintaining AltTab.

I've been trying to set this up, just to clarify my intended end-goal, I'm trying to mimic Windows (OS) window alt-tab behavior, and this is the final missing piece. What I have so far is:
1. Cmd+Tab => focus next window. (Mimics Alt-Tab in Windows)
2. Opt+Tab => Show all windows, after release do nothing. (Mimics Windows-Tab in Windows).

What I would like to be able to do is to see numbers on windows on Opt+Tab, honestly 0-9 is good enough for most use-cases, and seeing that this has been open for 5 years, I think we can start with working having numbers assigned to the first 10 windows, and later on decide how we'd like to handle the rest.

If we agree on this, I'll spend some time on it as soon as I have some free time.
What do you think? @lwouis

**Comment by lwouis** [comment]
Hi @AbdelrahmanHafez,

Thank you for your message. I think before going into implementation, what is required is specifications. I gave examples above:

* Do we hard-code the numbers 0 to 9, or do we let users customize? (I think hard-coding)
* How do we handle conflicts with other shortcuts? (e.g. the user has already bound `1` to some other action) (i think we may use the same approach we currently have for vim-keys)
* What about the ergonomics of numbers on the right of `4`? Is it really a good idea to map 0 to 9, given that 0 is the furthest and will require a 2 hand gesture? Maybe there are better alternatives? I'm not sure here. 0 to 9 seem un-ergonomic to me. I think at least we should not map 0 to the first window. At least start with 1.
* What about trackpad, do we react to it?
* Where to expose this new system in the Settings? Is it opt-in, or opt-out?
* What do the number indicators look like? How do we make them not confusing, if the other number indicators for Spaces are shown?
* Etc

<img width="486" height="306" alt="Image" src="https://github.com/user-attachments/assets/82658679-1253-4874-978d-666108898dcd" />

Thank you

**Comment by AbdelrahmanHafez** [comment]
Hello @lwouis

Happy to collaborate on this. I expect it'll be an iterative process rather than a one-shot implementation.

> Do we hard-code the numbers 0 to 9, or do we let users customize? (I think hard-coding)

Agreed. We can start with hard-coding and revisit customization later if there's demand.

> What about the ergonomics of numbers on the right of 4? Is it really a good idea to map 0 to 9, given that 0 is the furthest and will require a 2 hand gesture? Maybe there are better alternatives? I'm not sure here. 0 to 9 seem un-ergonomic to me. I think at least we should not map 0 to the first window. At least start with 1.

A few thoughts:
1. Agreed that 1 should select the first window, and 0 it should select 10th window.
2. I'd assume users of this feature already have both hands on the keyboard, so reaching for any number on the top row with the right hand shouldn't be problematic. Personally I use the numpad since it's more accurate.
3. I think of this as two separate one-handed gestures rather than a combined two-handed gesture. My typical usage would be: Opt+Tab to show windows in "sticky" mode (no action on release), then the left hand is free to press the desired number. Pressing Alt+Tab+Number simultaneously feels ergonomically awkward, so I'd expect most users to use this with sticky mode, or with numpad, if right hand is on the mouse then it's easier to click.

> What about trackpad, do we react to it?

I've never used AltTab with a trackpad, so I'm not sure how trackpad interaction currently works.

> What do the number indicators look like? How do we make them not confusing, if the other number indicators for Spaces are shown?

I found some interesting examples of clear button-style indicators. [This Google image search](https://www.google.com/search?q=keyboard+button+icon+with+number+in+it&sca_esv=4ba3af865432350d&rlz=1C5CHFA_enNL1061NL1061&udm=2&biw=2406&bih=1172&aic=0&sxsrf=ANbL-n668CBuM4KQ7-So9VZ3pS6QYJ0M1A%3A1768645665723&ei=IWRracXtK-qwi-gPssbfoQE&ved=0ahUKEwiFgo7FrpKSAxVq2AIHHTLjNxQQ4dUDCBI&uact=5&oq=keyboard+button+icon+with+number+in+it&gs_lp=Egtnd3Mtd2l6LWltZyIma2V5Ym9hcmQgYnV0dG9uIGljb24gd2l0aCBudW1iZXIgaW4gaXRIgxdQlAhYyQ5wAXgAkAEAmAE0oAGmA6oBATm4AQPIAQD4AQGYAgCgAgCYAwCIBgGSBwCgB5UDsgcAuAcAwgcAyAcAgAgA&sclient=gws-wiz-img#sv=CAMSVhoyKhBlLWxQN1hOQUttWFkwZDRNMg5sUDdYTkFLbVhZMGQ0TToOVUdJV3VEWU1kWVFFVU0gBCocCgZtb3NhaWMSEGUtbFA3WE5BS21YWTBkNE0YADABGAcg1sbkpAQwAkoKCAIQAhgCIAIoAg) for "keyboard button icon" shows some good options.

Another approach would be a semi-transparent overlay on each thumbnail with the number displayed. When users see sequential numbers across multiple windows, it should be intuitive that these are selection shortcuts rather than window counts.

> Where to expose this new system in the Settings? Is it opt-in, or opt-out?

I think it depends on the visual design. If the button style could be confused with window counts, it should be opt-in so users understand what they're enabling. If it's a distinct overlay design, opt-out would improve discoverability. We could also allow users to customize the display style.

> How do we handle conflicts with other shortcuts? (e.g. the user has already bound 1 to some other action) (i think we may use the same approach we currently have for vim-keys)

I don't think conflicts should be an issue here. This would be an implicit shortcut that only activates while the AltTab window is visible: "current shortcut to show AltTab" + number key. As far as I know, there's no way to bind a shortcut to just a number key without modifiers.


On a related note: I noticed that AltTab focuses the selected window with Space but not Enter. Is there a reason not to add Enter as well while working on this?

**Comment by lwouis** [comment]
Thank you for your message

It sounds good 👍  Here are a few comments:

> Agreed. We can start with hard-coding and revisit customization later if there's demand.

To be very clear, I was thinking to have a new toggle, similar to `vim keys` here:

<img width="672" height="773" alt="Image" src="https://github.com/user-attachments/assets/d5e7f5f7-bc79-40b9-8718-5d10f2e69d61" />

Note sure about if it would be opt-in or opt-out. It would depend on the UI for these direct shortcuts.

> I've never used AltTab with a trackpad, so I'm not sure how trackpad interaction currently works.

Sorry, I meant "numpad", not "trackpad". You mentioned you use it, so I assume it would be included. I actually don't know if keyboard numbers are merged/unified, or if numbers above the letters and numpad numbers are distinct input. This could complicate things a bit.

> Another approach would be a semi-transparent overlay on each thumbnail with the number displayed. When users see sequential numbers across multiple windows, it should be intuitive that these are selection shortcuts rather than window counts.

The issue is that some users already show the window Space. These are numbers, sitting in the corner. So we have to have a plan to include this in this feature. We can rework Space number if it helps. It's currently not the clearest UI in the first place, so I'm open to discussing what to do with it. The general point is that just a number on a thumbnail may not be clear. Is the window on monitor 2? On Space 2? Is it some order index? Is it a shortcut? We need to find something that mac users will understand to be a shortcut. The icon you linked seems a good start, with the cue that it's a keyboard key. It's not very mac-y though. Not sure what would be better though. Maybe we show the hold key + numbers, in the menubar item style?

<img width="227" height="85" alt="Image" src="https://github.com/user-attachments/assets/6179737c-830c-47d5-8208-cf7d3d644bc1" />

> I don't think conflicts should be an issue here. This would be an implicit shortcut that only activates while the AltTab window is visible: "current shortcut to show AltTab" + number key. As far as I know, there's no way to bind a shortcut to just a number key without modifiers.

It's possible (Settings > Controls) and some user probably do it since the numbers 1-to-4 are nicely located. AltTab also offer multiple triggers, so probably some people bind to alt+1, alt+2, alt+3 for the different modes. Examples of using the number 1:

<img width="672" height="773" alt="Image" src="https://github.com/user-attachments/assets/19955f2a-3434-4aee-91aa-d12b5c31bf53" />

<img width="672" height="773" alt="Image" src="https://github.com/user-attachments/assets/79058c12-ae2a-4c05-b861-4d685f730134" />

Thank you

**Comment by videni** [comment]
+1 for this feature. Simple number overlay on each thumbnail would be very intuitive for quick switching.

**Comment by wagnergerald** [comment]
Hi guys,

I am watching this quite a while now and made a small proof of concept for my desired workflow.
What do you think?

**TLDR: Small proof of concept when pressing and releasing ⌘ + ⇥ the window chooser is open, pressing the key right next to the app icon will bring the desired window to from. Quitting, closing, minimising modes are also available.**

**Demo**

Keystrokes are in the lower left corner

https://github.com/user-attachments/assets/56063d0d-5f85-4a38-bcba-f5e643ca4427

**The idea**

- Open the App with alt-Tab and let the chooser open
- The app has different modes, the default is chooser mode, quite like in Vimium, don't know where the idea is originally from
- Display letters right next to the app icon, which when pressed, apply the action of the current mode.
- Available Modes are: Choose, Quit, Close, Minimise

**The Fine print**

- Disclaimer, it's completely vibe coded and just for demonstration purposes. I did't look at the code at all
- Find the code here: https://github.com/wagnergerald/alt-tab-macos-vim-supercharged/tree/vimKeysSelection
- The app has a different bundle identifier and is named AltTabDev, so you can install and try it without modifying any AltTab App settings.
- There are also some bugs, where keys are not bound correctly, like at the end of the demo video.

**Comment by videni** [comment]
@wagnergerald    Good job,   could I ask why a straightforward number key isn’t used? I’d like to know if there are any technical challenges involved.I think a keyboard shortcut combination would be very intuitive — for example, press Alt+Tab, release Tab while holding Alt, then press any number key to switch.

**Comment by wagnergerald** [comment]
> could I ask why a straightforward number key isn’t used?

My considerations:
1) It's kinda like against my philosophy, leaving the "home row to reach for the numbers 😅
2) I saw this as my vibe-coding home project and it was more for learning as the outcome, after a while I got annoyed by the fact that adding new features broke new ones added before ...
3) No technical reason there, it should be easily doable.

**Comment by videni** [comment]
How would you handle 10+  applications using the HJKL keys?  Probably  an ephemeral number should suffice — much like Vim tab numbers.

**Comment by asselin** [comment]
There is an app on the app store called Switcheroo https://apps.apple.com/us/app/switcheroo-fast-app-switcher/id1198084875 that provides direct app access.  It's very simple-- you press right-Alt and it displays a list of running apps.  From there, you can customize the assignment of 0-9, a-z to each app.  When you hit right-Alt + the assigned letter/number, that app is directly switched to.  Unfortunately, that app is x64 only and the author isn't supporting it anymore, which is why I'd love to see something similar in Alttab.

Personally, I'd be satisfied with being able to assign keys within settings, e.g. similar to how alttab allows you to pick apps for blacklisting, it could allow you to pick an app and then assign a letter/number for a direct shortcut.
